### PR TITLE
rooted providers / event filtering

### PR DESCRIPTION
**Pull request by howufiger7**

### CODE REVIEW

File: cloudsync/event.py [hunk 1]
```diff
@@ -31,49 +30,72 @@ class Event:
     new_cursor: Optional[str] = None
 
 
+class ProviderGuard(set):
+    def add(self, provider):
+        if provider in self:
```
Comment (earonesty) [inline cloudsync/event.py:35]: we should consider adding a `clone()` function to providers.

Comment (howufiger7) [reply cloudsync/event.py:35]: that definitely crossed my mind as I was working on this piece -- decided on the simpler solution for now, will make `clone()` a separate task

File: cloudsync/tests/fixtures/mock_provider.py [hunk 2]
```diff
@@ -434,7 +477,8 @@ def _rename_single_object(self, source_object: MockFSObject, destination_path, *
         if self.oid_is_path:
             source_object.oid = destination_path
         self._store_object(source_object)
-        self._register_event(MockEvent.ACTION_RENAME, source_object, prior_oid)
+        if event:
```
Comment (reed-atakama) [inline cloudsync/tests/fixtures/mock_provider.py:480]: nit: Type annotation for `event`. Ambiguous if this a `bool` or an `Event`?

File: cloudsync/tests/test_cs.py [hunk 3]
```diff
@@ -106,15 +107,17 @@ def fixture_multi_remote_cs(mock_provider_generator):
     storage_dict: Dict[Any, Any] = dict()
```
Comment (reed-atakama) [inline cloudsync/tests/test_cs.py:107]: Update comment on line 103

File: cloudsync/event.py [hunk 4]
```diff
@@ -88,26 +109,34 @@ def forget(self):
...
-                        log.warning("Need auth, calling reauthenticate")
-                        try:
-                            self.reauthenticate()
-                        except NotImplementedError:
-                            raise CloudTokenError("No auth method defined")
-                    self.need_auth = False
-                else:
-                    log.info("reconnect to %s", self.provider.name)
-                    self.provider.reconnect()
+            self._reconnect()
```
[the remark this answers is not in the document]
Comment (earonesty) [reply cloudsync/event.py:139]: Or more simply, rename to "_reconnect_if_needed"

File: cloudsync/event.py [hunk 5]
```diff
@@ -127,16 +156,12 @@ def do(self):
...
+                      self.provider.name, self._root_path, self._root_oid)
             self._queue = []
             try:
-                if self.walk_root:
-                    for event in self.provider.walk(self.walk_root):
-                        self._process_event(event, from_walk=True)
-                else:
-                    for event in self.provider.walk_oid(self.walk_oid):
+                if self._root_oid:
+                    for event in self.provider.walk_oid(self._root_oid):
```
[the remark this answers is not in the document]
Comment (earonesty) [reply cloudsync/event.py:164]: technically, it should look it up... and use whatever's there....

File: cloudsync/event.py [hunk 6]
```diff
@@ -88,26 +109,34 @@ def forget(self):
...
+                    log.warning("Need auth, calling reauthenticate")
+                    try:
+                        self.reauthenticate()
+                    except NotImplementedError:
+                        raise CloudTokenError("No auth method defined")
+                self.need_auth = False
+            else:
+                log.info("reconnect to %s", self.provider.name)
+                self.provider.reconnect()
+            self._validate_root()
```
[the remark this answers is not in the document]
Comment (howufiger7) [reply cloudsync/event.py:135]: _validate_root() is an online operation / requires the provider to be connected. Technically the provider passed into an EventManager should always be connected, but we only check for connection_id in the constructor, and we have a test where a disconnected provider with a connection_id is used (test_reconn_after_disconn in test_cs.py)

Implemented lazy validation here to satisfy that test.

File: cloudsync/event.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (howufiger7) [reply cloudsync/event.py:139]: sure, I'll make that clearer

File: cloudsync/event.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (howufiger7) [reply cloudsync/event.py:164]: root path/oid validation ensures that both are set, even if only one is passed in initially, so there should be no need to check both here (unless someone is mucking with "private" state)

File: cloudsync/tests/fixtures/mock_provider.py (hunk 2 above)
Comment (howufiger7) [reply cloudsync/tests/fixtures/mock_provider.py:480]: yup, makes sense.

File: cloudsync/tests/test_cs.py (hunk 3 above)
Comment (howufiger7) [reply cloudsync/tests/test_cs.py:107]: will do

File: cloudsync/event.py [hunk 7]
```diff
@@ -31,49 +30,71 @@ class Event:
...
-        self.provider = provider
+                 root_path: Optional[str] = None, reauth: Callable[[], None] = None,
+                 root_oid: Optional[str] = None):
+        log.debug("provider %s, root %s", provider.name, root_path)
 
-        if not self.provider.connection_id:
+        if not provider.connection_id:
             raise ValueError("provider must be connected when starting the event manager")
 
+        self._provider_guard.add(provider)
```
[the remark this answers is not in the document]
Comment (howufiger7) [reply cloudsync/event.py:59]: do you mean we should put all the provider checks into a property?

File: cloudsync/event.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reed-atakama) [reply cloudsync/event.py:135]: Got it okay cool

**Pull request merged**